Mathlib/NumberTheory/JacobiSum/Basic.lean

```diff
@@ -0,0 +1,116 @@
+  simp only [mul_comm, sub_sub_cancel]
+
+/-- The Jacobi sum is compatible with ring homomorphisms. -/
+lemma jacobiSum_ringHomComp {R'' : Type*} [CommRing R''] (χ ψ : MulChar R R') (f : R' →+* R'') :
+    jacobiSum (χ.ringHomComp f) (ψ.ringHomComp f) = f (jacobiSum χ ψ) := by
+  simp only [jacobiSum, MulChar.ringHomComp, MulChar.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk,
+    map_sum, map_mul]
+
+end Def
+
+/-!
+### Jacobi sums over finite fields
+-/
+
+section FiniteField
+
+variable {F R : Type*} [Field F] [Fintype F] [DecidableEq F] [CommRing R]
+
+/-- The Jacobi sum of two multiplicative characters on a finite field `F` can be written
+as a sum over `F \ {0,1}`. -/
+lemma jacobiSum_eq_sum_sdiff (χ ψ : MulChar F R) :
+    jacobiSum χ ψ = ∑ x ∈ univ \ {0,1}, χ x * ψ (1 - x) := by
+  simp only [jacobiSum, subset_univ, sum_sdiff_eq_sub, mem_singleton, zero_ne_one,
+    not_false_eq_true, sum_insert, isUnit_iff_ne_zero, ne_eq, not_true_eq_false,
+    MulCharClass.map_nonunit, sub_zero, map_one, mul_one, sum_singleton, sub_self, mul_zero,
+    add_zero]
+
+private lemma jacobiSum_eq_aux (χ ψ : MulChar F R) :
+    jacobiSum χ ψ = ∑ x : F, χ x + ∑ x : F, ψ x - Fintype.card F +
+                      ∑ x ∈ univ \ {0, 1}, (χ x - 1) * (ψ (1 - x) - 1) := by
+  rw [jacobiSum]
+  conv =>
+    enter [1, 2, x]
+    rw [show ∀ x y : R, x * y = x + y - 1 + (x - 1) * (y - 1) by intros; ring]
+  rw [sum_add_distrib, sum_sub_distrib, sum_add_distrib, sum_one_sub_eq_sum,
+    Fintype.card_eq_sum_ones, Nat.cast_sum, Nat.cast_one, sum_sdiff_eq_sub (subset_univ _),
+    ← sub_zero (_ - _ + _), add_sub_assoc]
+  congr
+  rw [sum_pair zero_ne_one, sub_zero, ψ.map_one, χ.map_one, sub_self, mul_zero, zero_mul, add_zero]
+
+/-- If `1` is the trivial multiplicative character on a finite field `F`, then `J(1,1) = #F-2`. -/
+theorem jacobiSum_triv_triv : jacobiSum (1 : MulChar F R) 1 = Fintype.card F - 2 := by
+  rw [show 1 = MulChar.trivial F R from rfl, jacobiSum_eq_sum_sdiff]
+  have : ∀ x ∈ univ \ {0, 1}, (MulChar.trivial F R) x * (MulChar.trivial F R) (1 - x) = 1 := by
+    intros x hx
+    have hx' : IsUnit (x * (1 - x)) := by
+      simp only [mem_sdiff, mem_univ, mem_insert, mem_singleton, not_or, ← ne_eq, true_and] at hx
+      simp only [isUnit_iff_ne_zero]
+      exact mul_ne_zero hx.1 <| sub_ne_zero.mpr hx.2.symm
+    rw [← map_mul, MulChar.trivial_apply, if_pos hx']
+  calc ∑ x ∈ univ \ {0, 1}, (MulChar.trivial F R) x * (MulChar.trivial F R) (1 - x)
+  _ = ∑ _ ∈ univ \ {0, 1}, 1 := sum_congr rfl this
+  _ = Finset.card (univ \ {0, 1}) := (cast_card _).symm
+  _ = Fintype.card F - 2 := by
+    rw [card_sdiff (subset_univ _), card_univ, card_pair zero_ne_one]
+    obtain ⟨m, hm⟩ : ∃ m : ℕ, Fintype.card F = 1 + m + 1 :=
+      Nat.exists_eq_add_of_lt Fintype.one_lt_card
+    rw [show 1 + m + 1 = m + 2 by ring] at hm
+    simp only [hm, add_tsub_cancel_right (α := ℕ), Nat.cast_add, Nat.cast_ofNat,
+      add_sub_cancel_right]
```

Comment:
Maybe a bit hacky to abuse `lt/le` like this:
```suggestion
    obtain ⟨m, hm⟩ : ∃ m : ℕ, Fintype.card F = m + 2 :=
      Nat.exists_eq_add_of_le' Fintype.one_lt_card
    rw [hm, add_tsub_cancel_right]
    simp only [Nat.cast_add, Nat.cast_ofNat, add_sub_cancel_right]
```
I guess we should just have `Fintype.two_le_card`.

Reply:
Golfed this a bit further using `Nat.cast_sub <| Nat.add_one_le_of_lt Fintype.one_lt_card`.